File: 2024/01/Day01.py
#!/usr/bin/env python3
"""
<Problem description here>
"""
import sys
import re
import numpy as np

class Day01:
    def __init__(self):
        self.input = None

        self.lines = []
        self.a = []
        self.b = []
        
        self.ParseArgs()
        self.ParseInput()

    def ParseArgs(self, args=None):
        import argparse

        parser = argparse.ArgumentParser('Day01')
        parser.add_argument('input', nargs='?', default='input')

        parser.parse_args(args, self)


    def ParseInput(self):
        with open(self.input) as input:
            self.lines = input.read().strip().split('\n')

        for line in self.lines:
            itemA, itemB = line.split()
            self.a.append(int(itemA))
            self.b.append(int(itemB))

    def Part1(self):
        self.a.sort()
        self.b.sort()

        answer = sum(abs(itemA - itemB) for itemA, itemB in zip(self.a, self.b))
        return answer

    def Part2(self):
        answer = sum(itemA * self.b.count(itemA) for itemA in self.a)
        return answer
    
if __name__ == '__main__':
    problem = Day01()
    
    answer1 = problem.Part1()
    print(f'Answer 1: {answer1}')

    answer2 = problem.Part2()
    print(f'Answer 2: {answer2}')

File: 2024/01/Day01.cpp
#include <iostream>
#include <fstream>
#include <set>
#include <cassert>

class Day01
{
public:
    Day01(std::string name) {
        GetInput(name);
    }

    void GetInput(std::string name) {
        std::ifstream input{name};
        assert(not input.fail());

        int a, b;

        while (not input.eof()) {
            input >> a >> b;
            if (input.fail()) {
                break;
            }
            listA.insert(a);
            listB.insert(b);
        }
    }

    unsigned long Part1() {
        unsigned long answer = 0;

        for (auto iterA = listA.begin(), iterB=listB.begin();
             iterA != listA.end() && iterB != listB.end();
             ++iterA, ++iterB)
        {
            answer += abs(*iterA - *iterB);
        }

        return answer;
    }

    unsigned long Part2() {
        unsigned long answer = 0;

        for (auto iterA = listA.begin(); iterA != listA.end(); ++iterA) {
            answer += *iterA * listB.count(*iterA);
        }

        return answer;
    }

    std::multiset<int> listA;
    std::multiset<int> listB;

    std::string inputName;
};

int main(int argc, char** argv)
{
    std::string inputName(argc > 1 ? argv[1] : "input");
    Day01 problem(inputName);

    std::cout << problem.Part1() << std::endl;
    std::cout << problem.Part2() << std::endl;

    return 0;
}